controllers/common/fetch.go

```diff
@@ -115,57 +119,69 @@ func FetchSubscription(c client.Client, name, namespace string, packageName ...s
-		return nil, nil
-	}
-
-	csvName := sub.Status.CurrentCSV
+	csvName := ""
 	csvNamespace := sub.Namespace
 
-	if sub.Status.Install == nil || sub.Status.InstallPlanRef.Name == "" {
+	if sub.Status.InstallPlanRef != nil {
+		ip, err := FetchInstallPlan(c, sub.Status.InstallPlanRef.Name, sub.Status.InstallPlanRef.Namespace)
+		if err != nil {
+			return nil, err
+		}
+		for _, step := range ip.Status.Plan {
+			if step.Resource.Kind == "ClusterServiceVersion" && strings.HasPrefix(step.Resource.Name, sub.Name) {
+				if step.Status == olmv1alpha1.StepStatusPresent {
+					csvName = step.Resource.Name
+					break
+				}
+				klog.Warningf("ClusterServiceVersion %s is not present", step.Resource.Name)
+				return nil, nil
+			}
+		}
+	} else {
 		klog.Warningf("The Installplan for Subscription %s is not ready. Will check it again", sub.Name)
 		return nil, nil
 	}
 
-	// If the installplan is deleted after is completed, ODLM won't block the CR update.
-
-	ipName := sub.Status.InstallPlanRef.Name
-	ipNamespace := sub.Namespace
-	ip := &olmv1alpha1.InstallPlan{}
-	ipKey := types.NamespacedName{
-		Name:      ipName,
-		Namespace: ipNamespace,
-	}
-	if err := c.Get(context.TODO(), ipKey, ip); err != nil && !errors.IsNotFound(err) {
-		klog.Errorf("failed to get Installplan %s in the namespace %s: %s", ipName, ipNamespace, err)
-		return nil, err
-	} else if !errors.IsNotFound(err) && ip.Status.Phase == olmv1alpha1.InstallPlanPhaseFailed {
-		klog.Errorf("installplan %s in the namespace %s is failed", ipName, ipNamespace)
-	} else if !errors.IsNotFound(err) && ip.Status.Phase != olmv1alpha1.InstallPlanPhaseComplete {
-		klog.Warningf("Installplan %s in the namespace %s is not ready", ipName, ipNamespace)
-		return nil, nil
-	}
-
 	csv := &olmv1alpha1.ClusterServiceVersion{}
 	csvKey := types.NamespacedName{
 		Name:      csvName,
 		Namespace: csvNamespace,
 	}
-	if err := c.Get(context.TODO(), csvKey, csv); err != nil {
-		if errors.IsNotFound(err) {
-			klog.V(3).Infof("ClusterServiceVersion %s is not ready. Will check it when it is stable", sub.Name)
-			return nil, nil
-		}
-		klog.Errorf("failed to get ClusterServiceVersion %s in the namespace %s: %s", csvName, csvNamespace, err)
+	if err := getObjectWithRetry(c, csvKey, csv); err != nil {
 		return nil, err
 	}
-
-	klog.V(3).Infof("Get ClusterServiceVersion %s in the namespace %s", csvName, csvNamespace)
 	return csv, nil
 }
 
+func FetchInstallPlan(c client.Client, name, namespace string) (*olmv1alpha1.InstallPlan, error) {
+	installPlan := &olmv1alpha1.InstallPlan{}
+	installPlanKey := types.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	}
+	if err := getObjectWithRetry(c, installPlanKey, installPlan); err != nil {
+		return nil, err
+	}
+	return installPlan, nil
+}
+
+// Get object with retry
+func getObjectWithRetry(c client.Client, key types.NamespacedName, obj runtime.Object) error {
+	if err := wait.PollImmediate(time.Second*20, time.Minute*1, func() (bool, error) {
```

Comment:
@DanielXLee I am considering if waiting for one minute is too long. How about `PollImmediate(time.Second*2, time.Second*20,`?